libnvme/src/firmware.rs

```diff
@@ -0,0 +1,306 @@
+impl TryFrom<u8> for NvmeSlot {
+    type Error = NvmeSlotError;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            1..=7 => Ok(NvmeSlot(value as _)),
+            invalid => Err(NvmeSlotError::InvalidSlotNumber(invalid as _)),
+        }
+    }
+}
+
+#[derive(Debug)]
+/// The Firmwware log page for an NVMe controller.
+pub struct FirmwareLogPage {
+    /// The currently active firmware slot.
+    pub active_slot: u8,
+    /// The next slot that will be active on a controller reset.
+    pub next_active_slot: Option<u8>,
+    /// Slot 1 is read-only.
+    pub slot1_is_read_only: bool,
+    /// The number of firmware slots the controller has.
+    pub number_of_slots: usize,
+    firmware_slots: Vec<Option<String>>,
+}
+
+impl FirmwareLogPage {
+    fn init(
+        identify: &ControllerInfoIdentify<'_>,
+        logpage: nvme_fwslot_log_t,
+    ) -> Self {
+        let active_slot = logpage.bitfield1.fw_afi();
+        let slot1_is_read_only =
+            unsafe { (*identify.inner).id_frmw.fw_readonly() };
+
+        // NVMe Spec: "If this field is 0h, then the controller does not
+        // indicate the firmware slot that is going to be activated at the next
+        // Controller Level Reset."
+        let next_active_slot = match logpage.bitfield1.fw_next() {
+            0 => None,
+            slot => Some(slot.saturating_sub(1)),
+        };
+
+        let number_of_slots =
+            unsafe { (*identify.inner).id_frmw.fw_nslot() } as usize;
+        let mut firmware_slots = Vec::with_capacity(number_of_slots);
+        for slot in &logpage.fw_frs[..number_of_slots] {
+            // FIXME there's probably a better way to do this? Basically the
+            // strings are packed into an array without a nul byte unless the
+            // slot itself is empty so we can't depend on `CStr::from_ptr()`.
+            // Instead we have to take each slice of bytes and convert it into a
+            // CString checking if the first byte is nul along the way.
+            let bytes: Vec<u8> = slot.iter().map(|&x| x as u8).collect();
+            if bytes[0] == b'\0' {
+                firmware_slots.push(None);
+                continue;
+            }
+            let cstring = unsafe { CString::from_vec_unchecked(bytes) };
+            // NVMe Spec: "The firmware revision is indicated as an ASCII
+            // string."
+            let firmware = cstring.to_string_lossy().trim().to_string();
+            firmware_slots.push(Some(firmware));
+        }
+
+        Self {
+            active_slot,
+            next_active_slot,
+            slot1_is_read_only,
+            number_of_slots,
+            firmware_slots,
+        }
+    }
+
+    /// Get the firmware version for a particular slot.
+    ///
+    /// Note that the NVMe spec allows for slots 1 - 7.
+    pub fn get_slot<S>(&self, slot: S) -> Result<Option<&String>, NvmeSlotError>
```

Comment:
Same naming comment as above - maybe `get_slot_version`? (Also strongly agree with @pfmooney's comment - I'd have this just take `slot: NvmeSlot` here and skip the generics, especially because that allows you to return just `Option<_>` instead of `Result<Option<_>, _>`, I think?)

Small nit: `Option<&String>` is a little weird - maybe return `Option<&str>` instead via `.as_deref()` instead of `.as_ref()`

Reply:
Fixed in dff4996b